bin/ubuntu-core-initramfs

```diff
@@ -486,6 +486,139 @@ def install_misc(dest_dir, sysroot):
     # FIXME: systemd is configured with the wrong path to dmsetup
     os.symlink("../usr/sbin/dmsetup", os.path.join(dest_dir, "sbin", "dmsetup"))
 
+
+class AptRepo:
+
+    def __init__(self, url):
+        self.url = url
+        self.suites = []
+        self.components = []
+
+
+def find_apt_repos(sysroot):
+    repos = {}
+    sources = os.path.join(sysroot, "etc/apt/sources.list")
+    sources_d = os.path.join(sysroot, "etc/apt/sources.list.d")
+    state_d = os.path.join(sysroot, "var/lib/apt/lists")
+    repo_lines = check_output(["apt-cache", "policy",
+                               "-o", "Dir::Etc::SourceList="+sources,
+                               "-o", "Dir::Etc::SourceParts="+sources_d,
+                               "-o", "Dir::State::Lists="+state_d]).decode("utf-8").splitlines()
+    uri_line = re.compile(r"^ *-?[0-9]+ (http|mirror|file|cdrom|ftp|copy|rsh|ssh)")
+    repo_lines = [r for r in repo_lines if uri_line.match(r)]
+    uri_start = re.compile(r"^ *-?[0-9]+ ")
+    for r in repo_lines:
+        r = uri_start.sub("", r)
+        fields = r.split(" ")
+        if len(fields) < 2:
+            continue
+        # suite/comp in the line
+        suite_comp = fields[1].split("/")
+        if len(suite_comp) != 2:
+            continue
+        # Key will be url + suite (we only have one suite per line)
+        key = (fields[0], suite_comp[0])
+        if key in repos:
+            repos[key].components.append(suite_comp[1])
+        else:
+            repo = AptRepo(fields[0])
+            repo.suites = [suite_comp[0]]
+            repo.components = [suite_comp[1]]
+            repos[key] = repo
+    # Now merge suites that share components
+    repo_ls = list(repos.values())
+    merged = set()
+    final_repos = []
+    for i, repo in enumerate(repo_ls):
+        if i in merged:
+            continue
+        for j in range(i+1, len(repo_ls)):
+            if sorted(repo.components) == sorted(repo_ls[j].components):
+                repo.suites += repo_ls[j].suites
+                merged.add(j)
+        final_repos.append(repo)
+
+    return final_repos
+
+
+def pkgs_used_in_initrd(dest_dir, sysroot):
+    MD5_LEN = 32
```

Comment:
As I did not want to use hashlib, I did not want to add the dependency just because of the constant. But good point on naming, I have changed it no MD5_HEX_CHARS.